prov/efa/src/rxr/rxr_pkt_type_misc.c

```diff
@@ -63,45 +63,58 @@ ssize_t rxr_pkt_init_handshake(struct rxr_ep *ep,
 
 	pkt_entry->pkt_size = sizeof(struct rxr_handshake_hdr)
 			      + RXR_NUM_PROTOCOL_VERSION * sizeof(uint64_t);
-	pkt_entry->addr = addr;
+	pkt_entry->addr = tx_entry->addr;
+	pkt_entry->x_entry = tx_entry;
 	return 0;
 }
 
-void rxr_pkt_post_handshake(struct rxr_ep *ep,
-			    struct rxr_peer *peer,
-			    fi_addr_t addr)
+ssize_t rxr_pkt_post_handshake(struct rxr_ep *ep, struct rxr_peer *peer,
+				fi_addr_t addr)
 {
-	struct rxr_pkt_entry *pkt_entry;
+	struct rxr_tx_entry *tx_entry;
 	ssize_t ret;
 
 	assert(!(peer->flags & RXR_PEER_HANDSHAKE_SENT));
 
-	pkt_entry = rxr_pkt_entry_alloc(ep, ep->tx_pkt_efa_pool);
-	if (OFI_UNLIKELY(!pkt_entry))
-		return;
-
-	rxr_pkt_init_handshake(ep, pkt_entry, addr);
-
-	/*
-	 * TODO: Once we start using a core's selective completion capability,
-	 * post the HANDSHAKE packets without FI_COMPLETION.
-	 */
-	ret = rxr_pkt_entry_send(ep, pkt_entry, addr);
+	tx_entry = ofi_buf_alloc(ep->tx_entry_pool);
+	if (OFI_UNLIKELY(!tx_entry)) {
+		FI_WARN(&rxr_prov, FI_LOG_EP_CTRL, "TX entries exhausted.\n");
+		return -FI_EAGAIN;
```

Comment:
-FI_ENOMEM is more appropriate. The caller cannot retry.

Reply:
Now we queue the peer, so return -FI_EAGAIN in order to retry.